pkg/controllers/machinemigration/machine_migration_controller.go

```diff
@@ -133,12 +134,16 @@ func (r *MachineMigrationReconciler) Reconcile(ctx context.Context, req reconcil
 		return ctrl.Result{}, nil
 	}
 
-	// Check if the Synchronized condition is set to True.
-	// If it is not, this indicates an unmigratable resource and therefore should take no action.
-	if cond, err := util.GetConditionStatus(mapiMachine, string(controllers.SynchronizedCondition)); err != nil {
-		return ctrl.Result{}, fmt.Errorf("unable to get synchronizedCondition for %s/%s: %w", mapiMachine.Namespace, mapiMachine.Name, err)
-	} else if cond != corev1.ConditionTrue {
-		logger.Info("Machine not synchronized with latest authoritative generation, will retry later")
+	// Check that the resource is synchronized and up-to-date.
+	//
+	// This MUST be checked BEFORE setting status.authoritativeAPI to Migrating,
+	// because after that the sync controller will not run to update it and we
+	// will deadlock.
+	if isSynchronized, err := r.isSynchronized(ctx, mapiMachine); err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to check synchronizedGeneration matches old authority: %w", err)
```

Comment:
The semantics of isSynchornized cover more ground now
```suggestion
		return ctrl.Result{}, fmt.Errorf("unable to check the resource is synchronized and up-to-date with its authority: %w", err)
```